### DISCUSSION TITLE
Middleware wrapping call with AsyncLocalStorage.run()

### DISCUSSION THREAD

**Discussion by crummy** [RobinTail/express-zod-api#1650] [Q&A] [answered]
The moment a request comes in, I'd like to generate a traceId, which I'll include in any logs my application generates. Here's how I do this with a different API library (TRPC):

```
const traceIdStorage = new AsyncLocalStorage<string>();
export const internalProcedure = internal.procedure.use(({ next }) => {
    return traceIdStorage.run(randomUUID(), async () => {
      const result = await next({ ctx });
      if (!result.ok) {
        logger.error("Something went wrong", result.error);
      }
      return result;
    });
  },
);
```

I set up my logger to read from traceIdStorage so I can include that data in any logs I send.

I'd like to do the same with express-zod-api, but the middlewares are a little different - there's no `next` object I can call in my `.run()` argument. Is there another way to achieve this?

**Comment by crummy** [comment]
Think I've found it:

```
const config = createConfig({
  server: {
    listen: 8090, // port or socket
    compression: true,
    beforeRouting: ({ app }) => {
      app.use(({next}) => {
        return traceId.run(randomUUID(), async () => {
          try {
            if (next) next()
          } catch (e) {
            console.error("Uncaught exception", e)
            throw e;
          }
        })
      }
    }
  },
}
```

However, it would be handy to be able to do it later, where I have figured out who my user is (then I can attach usernames etc to my logger context)

**Reply by crummy** [reply]
This doesn't really work it turns out - it seems I do not have access to the traceId within my code, perhaps `next` is not being called as I expected, or something else is going on.

**Reply by RobinTail** [reply]
if you wanna go this way, `next` should be the third argument of `app.use`

**Reply by crummy** [reply]
That worked great too! @RobinTail thanks so much for your hard work and support!

**Comment by RobinTail** [comment, marked as answer]
Hello, @crummy 

There is a dedicated article on that topic:
https://github.com/RobinTail/express-zod-api?tab=readme-ov-file#child-logger

Implying that your logger has a `.child()` method:

```ts
import { createConfig } from "express-zod-api";
import { randomUUID } from "node:crypto";

const config = createConfig({
  // logger: ...,
  childLoggerProvider: ({ parent, request }) =>
    parent.child({ requestId: randomUUID() }),
});
```

Basically you simply create another instance of the logger inside that function, that somehow displays the request ID.
So that you don't even need the storage — this approach enables a dedicated instance of the logger for the individual request.
I hope it helps.

**Reply by crummy** [reply]
Thank you! Does that setup work with TypeScript, or do I need to cast perhaps? There is no `child` field on the `parent` logger.

**Reply by RobinTail** [reply]
It depends on your actual logger. If you're using the default one (winston) or pino, then they DO have `.child` method, BUT you'd need to use module augmentation technique to declare it:

```ts
// Setting the type of the logger used — somewhere in your code (e.g. near config)
declare module "express-zod-api" {
  interface LoggerOverrides extends Logger {} // import Logger from your actual logger
}
```

After doing that, every `logger` will be fully typed as the actual logger you use.

If your actual logger does NOT have such method, you'd need to make an implementation for `childLoggerProvider` that returns an object compatible with `AbstractLogger` (having 4 logging methods), @crummy 

```ts
const config = createConfig({
  // logger: ...,
  childLoggerProvider: ({ parent, request }) => ({
    debug: (msg, extra) => parent.debug(msg, {...extra, requestId: randomUUID() }),
    // info, warn, error, 
  })
});
```

**Reply by crummy** [reply]
Thanks so much!

The one thing I'm missing, that my other solution provides, is the ability to retrieve the UUID after this (e.g. to provide a traceid header to a fetch request later on). Is there a way to do this with the `childLoggerProvider`? It seems winston doesn't support retrieving these fields after the fact, but I noticed the `request` object has a `next` field - is it possible to use this with the `asyncLocalStorage` technique to store + wrap the request + retrieve?

**Reply by RobinTail** [reply]
You can use `request.res.locals` for storage, @crummy :
https://expressjs.com/en/4x/api.html#res.locals

I can also make the `response` available directly in `childLoggerProvider` so that you'd not need to use the `req.res` reference.

That can also simplify setting a header to response immediately if you mean that you want the `requestId` to be in a header of response. Tell me if you need `response` as an argument of `childLoggerProvider`, @crummy

**Reply by crummy** [reply]
Thanks! That's a little less elegant than the AsyncLocalStorage but it works fine enough for me.

For posterity, here is how I set up my logger:

```
  childLoggerProvider: ({ parent, request }) => {
    const traceId = randomUUID();
    if (request.res) {
      request.res.locals = { traceId };
    }
    return parent.child({ traceId, path: request.path });
  },
```
Later I have a middleware that extracts the traceId:
```
   createMiddleware({
      const traceId = request.res?.locals.traceId;
      return { traceId }
  })
```

And then I use this to retrieve it:
```
  handler: async ({ input, logger, options }) => {
    const traceId = options.traceId;
  }
```

**Reply by RobinTail** [reply]
Wonderful.
- I will add `response` to `childLoggerProvider`, so that you'd not need to refer to it via `request.res`, @crummy .
- In the `createMiddleware` you can take it directly from `response` which is already available there.

Could you clarify one more thing for me regarding `traceId` in the endpoint handler:

> a traceId header to a fetch request later on

If that the `fetch` from some other external data source needed for your API to respond?
Like your API is using another API — is that correct?
Because if so, I have one more idea, but I need to ensure that I understand the use-case correctly.

**Reply by crummy** [reply]
Yes, my express-zod-api service sometimes makes HTTP calls to other services we maintain. Ideally we generate a traceId the moment we receive a public request, and then hand that traceId to any other services, who pass it on, etc. 

So later on I make a request like:
```
fetch(`${internalService}/api/whatever`, { headers: { 'trace-id': asyncLocalStorage.getStore() } }
```

**Reply by RobinTail** [reply]
Ok. While I'm thinking on a better approach, here is one more suggestion on excluding `request.res.locals` from that:
Some sources state that it's actually a normal practice to add custom properties to request, like this:

```ts
declare global {
  namespace Express {
    interface Request {
      traceId?: string;
    }
  }
}
```

https://dev.to/kwabenberko/extend-express-s-request-object-with-typescript-declaration-merging-1nn5
https://stackoverflow.com/questions/37377731/extend-express-request-object-using-typescript

So in `childLoggerProvider` you can set `request.traceId` and then promote it as an option for endpoints using middleware, but taking it from request — that would probably improve clarify and semantics, @crummy

**Reply by RobinTail** [reply]
Update on this topic.
Starting v19.2.0 `.child()` method added to the built-in logger
